### PR TITLE
Switch to create-dmg

### PR DESCRIPTION
**Pull request by ann0see**
**Short description of changes**

Just a draft on how we could switch to create-dmg. Probably it's worth changing more in the macOS build process.
**Context: Fixes an issue?**

Fixes: #1797 
**Does this change need documentation? What needs to be documented and how?**
No

**Status of this Pull Request**

Working semi-tested version 
**What is missing until this pull request can be merged?**
Someone with a Mac should test if it does what it should. Especially the placing of files in the .dmg
And it should be squashed and merged
## Checklist

- [x] I've verified that this Pull Request follows the [general code principles](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)
- [x] I tested my code and it does what I want
- [x] My code follows the [style guide](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#source-code-consistency) 
- [x] I waited some time after this Pull Request was opened and all GitHub checks completed without errors. 
- [x] I've filled all the content above

### CODE REVIEW

File: .github/workflows/autobuild.yml
Comment (ann0see) [inline .github/workflows/autobuild.yml]:
```suggestion
             building_on_os:        windows-latest
```

**Comment by ann0see** [comment]
This should NOT be merged as is.

**Review by ann0see — COMMENTED** [review]
The .DS_Store file doesn't seem to be created.

**Comment by ann0see** [comment]
@emlynmac could you please Test if this build works on a modern Mac? I couldn’t really test it on my macOS VM…

**Comment by emlynmac** [comment]
> @emlynmac could you please test if this build works on a modern Mac? I couldn’t really test it on my macOS VM…

Sure - where can I find the build?

**Comment by ann0see** [comment]
The .dmg is in the actions tab as artifact; however I‘m more interested in a local creation of the .dmg file. On GH actions it seems as if it doesn’t correct create a .DS_Store file and therefore doesn’t add a background image

**Comment by emlynmac** [comment]
> The .dmg is in the actions tab as artifact; however I‘m more interested in a local creation of the .dmg file. On GH actions it seems as if it doesn’t correct create a .DS_Store file and therefore doesn’t add a background image

I'd not be concerned about the .DS_Store. It's just a part of the filesystem.
The image should be a different file somewhere.
The DMG opened fine on my 12.0.1 macPro.

**Comment by ann0see** [comment]
I think the background-image file is included in the dmg (see hidden folders).

File: .github/workflows/autobuild.yml
Comment (softins) [inline .github/workflows/autobuild.yml:161]: Note that although the MacOS build appeared to succeed, it issued the following warnings:
```
Project WARNING: Qt has only been tested with version 10.15 of the platform SDK, you're using 12.0.
Project WARNING: This is an unsupported configuration. You may experience build issues, and by using
Project WARNING: the 12.0 SDK you are opting in to new features that Qt has not been prepared for.
Project WARNING: Please downgrade the SDK you use to build your app to version 10.15, or configure
Project WARNING: with CONFIG+=sdk_no_version_check when running qmake to silence this warning.
```
That is the reason why selecting the SDK version was added.

I expect it applies even more to the legacy build. Although it does not appear to do the above check, the warning probably still applies.

Comment (ann0see) [reply .github/workflows/autobuild.yml:161]: Ok. So leave it?

File: mac/deploy_mac.sh
Comment (pljones) [inline mac/deploy_mac.sh:76]: Could you wrap this line nicely, so Github diff doesn't scroll sideways, please...  (clang-format for shell scripts needed... mmm...)

Comment (ann0see) [reply mac/deploy_mac.sh]: Done. I hope it still works (there were errors in the script as I re-ordered the arguments)

File: mac/deploy_mac.sh
Comment (pljones) [inline mac/deploy_mac.sh:85]: Thanks!

**Comment by ann0see** [comment]
I seem to get what the problem is now... We need to copy the .DS_Store we already had so far to the new installer since CI doesn't like GUI stuff (and we don't get permissions to execute a apple script create-dmg uses to set the background images)

**Comment by ann0see** [comment]
... or we use a different tool.

**Comment by ann0see** [comment]
Nope. Copying the .DS_Store also doesn't work...

File: mac/deploy_mac.sh
Comment (pljones) [inline mac/deploy_mac.sh:77]: What is this?  16KB binary file?

Comment (ann0see) [reply mac/deploy_mac.sh:77]: It's the .DS_store file from the old installer. Unfortunately it gets overwritten. I don't really think this create-dmg script is the correct tool for us (much too hacky). Maybe we should use a node-js version instead: https://github.com/sindresorhus/create-dmg

Comment (ann0see) [reply mac/deploy_mac.sh:77]: Or this one: https://github.com/LinusU/node-appdmg

Comment (ann0see) [reply mac/deploy_mac.sh:77]: Just found https://github.com/actions/virtual-environments/issues/553

Comment (ann0see) [reply mac/deploy_mac.sh]: Urgh. Certainly overthought the whole stuff. It seems to work with GH actions!!!!

**Comment by ann0see** [comment]
@emlynmac @pljones ready to review.

File: mac/deploy_mac.sh
Comment (ann0see) [inline mac/deploy_mac.sh:72]: Do we want to rely on brew?

**Comment by ann0see** [comment]
It even seems to fix the scroll bars being shown on Big Sur upwards

**Review by pljones — APPROVED** [review]

**Comment by ann0see** [comment]
@softins I‘d like to have this in the next release too.

**Comment by softins** [comment]
> @softins I‘d like to have this in the next release too.

I would prefer we stick to what we know and have already for 3.8.2, and review this properly for use in 3.9.0.

Is there anything that makes it urgent for 3.8.2?

**Comment by softins** [comment]
I'm particularly concerned about the switch to macos-11 and the newer SDK, since Qt 5.15.2 explicitly gives warnings about using that SDK. I would rather have more time to evaluate that without delaying 3.8.2 to do so.

**Comment by ann0see** [comment]
Ok. This is no longer using an updated SDK btw.

**Comment by ann0see** [comment]
Sorry to say that, but I am no longer that confident that the scroll bar issue is fixed. I just realised that the Mac I tested the dmg on was running some OS before big sur.

File: mac/deploy_mac.sh
Comment (hoffie) [reply mac/deploy_mac.sh:72]: Is there a better alternative?
I think we should keep external package manager usage to a minimum, but using one is certainly better than download binaries (and their dependencies!) manually all over the place.

If we do use brew, we should update #2277 accordingly.

Can you do version pinning here?

**Comment by hoffie** [comment]
> This should NOT be merged as is.

Marking as Draft per that comment.

**Comment by hoffie** [comment]
I've had a quick look at the .dmg build artifact and noticed that it seems to be a different format. I can't yet tell what's better or if it makes a difference at all, but I wanted to mention it.

<details>
<summary>7z x ../jamulus_3.8.1dev-6172bc8_mac.dmg (this PR)</summary>

```
$ 7z x ../jamulus_3.8.1dev-6172bc8_mac.dmg

7-Zip [64] 17.04 : Copyright (c) 1999-2021 Igor Pavlov : 2017-08-28
p7zip Version 17.04 (locale=de_DE.UTF-8,Utf16=on,HugeFiles=on,64 bits,8 CPUs x64)

Scanning the drive for archives:
1 file, 24495216 bytes (24 MiB)

Extracting archive: ../jamulus_3.8.1dev-6172bc8_mac.dmg
--
Path = ../jamulus_3.8.1dev-6172bc8_mac.dmg
Type = Dmg
Physical Size = 24495216
Method = Zero2 ZLIB CRC
Blocks = 68
----
Path = 4.hfs
Size = 85983232
Packed Size = 24452455
Comment = disk image (Apple_HFS : 4)
Method = Zero2 ZLIB CRC
--
Path = 4.hfs
Type = HFS
Physical Size = 85983232
Method = HFS+
Cluster Size = 4096
Free Space = 27480064
Created = 2022-01-18 09:57:57
Modified = 2022-01-18 10:58:31

Everything is Ok                                               

Folders: 108
Files: 167
Size:       57592322
Compressed: 24495216
```
</details>


<details>
<summary>7z x ../jamulus_3.8.1dev-b324ee8_mac.dmg (master)</summary>

```
$ 7z x ../jamulus_3.8.1dev-b324ee8_mac.dmg # master

7-Zip [64] 17.04 : Copyright (c) 1999-2021 Igor Pavlov : 2017-08-28
p7zip Version 17.04 (locale=de_DE.UTF-8,Utf16=on,HugeFiles=on,64 bits,8 CPUs x64)

Scanning the drive for archives:
1 file, 23274493 bytes (23 MiB)

Extracting archive: ../jamulus_3.8.1dev-b324ee8_mac.dmg
WARNING:
../jamulus_3.8.1dev-b324ee8_mac.dmg
Can not open the file as [Dmg] archive
The file is open as [bzip2] archive

--
Path = ../jamulus_3.8.1dev-b324ee8_mac.dmg
Open WARNING: Can not open the file as [Dmg] archive
Type = bzip2

ERROR: There are some data after the end of the payload data : jamulus_3.8.1dev-b324ee8_mac

Sub items Errors: 1

Archives with Errors: 1

Sub items Errors: 1
```
</details>

File: mac/deploy_mac.sh
Comment (ann0see) [reply mac/deploy_mac.sh]: Not sure how. I read it should work with @version but the ci failed

Comment (hoffie) [reply mac/deploy_mac.sh]: Looks like there is no easy way. :(

This seems to work but sounds strange to me. It also downloads a rather huge git repo to do that.
```
root@cc5850aba45c:/home/linuxbrew# brew extract --version=1.0.9 create-dmg homebrew/cask
Warning: extract is a developer command, so
Homebrew's developer mode has been automatically turned on.
To turn developer mode off, run brew developer off

==> Tapping homebrew/cask
Cloning into '/home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-cask'...
remote: Enumerating objects: 618845, done.
remote: Total 618845 (delta 0), reused 0 (delta 0), pack-reused 618845
Receiving objects: 100% (618845/618845), 282.97 MiB | 7.00 MiB/s, done.
Resolving deltas: 100% (437655/437655), done.
Tapped 3959 casks (4,030 files, 302.8MB).
==> Searching repository history
==> Writing formula for create-dmg from revision 61daf01 to:
/home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-cask/Formula/create-dmg@1.0.9.rb
root@cc5850aba45c:/home/linuxbrew# brew install /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-cask/Formula/create-dmg@1.0.9.rb
Error: Failed to load cask: /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-cask/Formula/create-dmg@1.0.9.rb
Cask 'create-dmg@1.0.9' is unreadable: wrong constant name #<Class:0x000000000270c540>
Warning: Treating /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-cask/Formula/create-dmg@1.0.9.rb as a formula.
==> Downloading https://github.com/create-dmg/create-dmg/archive/v1.0.9.tar.gz
==> Downloading from https://codeload.github.com/create-dmg/create-dmg/tar.gz/v1.0.9
#=#=#                                                                         
==> Installing create-dmg@1.0.9 from homebrew/cask
==> make install prefix=/home/linuxbrew/.linuxbrew/Cellar/create-dmg@1.0.9/1.0.9
🍺  /home/linuxbrew/.linuxbrew/Cellar/create-dmg@1.0.9/1.0.9: 12 files, 61.4KB, built in 1 second
==> Running `brew cleanup create-dmg@1.0.9`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
```

Comment (ann0see) [reply mac/deploy_mac.sh:75]: Didn't seem to work on the CI. I've reverted it.

Comment (ann0see) [reply mac/deploy_mac.sh:75]: Argh. You need an additional install step

**Comment by ann0see** [comment]
Should be tested once again and then squash merged. There are no SDK changes anymore

**Comment by hoffie** [comment]
> I've had a quick look at the .dmg build artifact and noticed that it seems to be a different format. I can't yet tell what's better or if it makes a difference at all, but I wanted to mention it.

I think someone with macOS knowledge should judge what this means. To me it just means that this PR does not only change the process, but it also changes the result, either as a result of the different tooling or as a result of a different build environment (hdituil?). This change might be positive for us and for users or it might unexpectedly break for older macOS releases. I can't tell, but someone should. :)

File: mac/deploy_mac.sh
Comment (hoffie) [inline mac/deploy_mac.sh:73]:
```suggestion
    local create_dmg_version="1.0.9"
    brew extract --version="${create_dmg_version}" create-dmg homebrew/cask
    brew install /usr/local/Homebrew/Library/Taps/homebrew/homebrew-cask/Formula/create-dmg@"${create_dmg_version}".rb
```

**Review by hoffie — COMMENTED** [review]
Another nit. Happy to approve once there's confirmation regarding the file format.
This is probably the explanation for the format within the old build, but we still don't know why this format was chosen:
https://github.com/jamulussoftware/jamulus/pull/2207/files#diff-8f016f4f41c7ad708e4ff8d2546ce6fc8a4e3146159b2a621688c472f912ff7cL38

According to git history, this was introduced along the initial mac build logic by @doloopuntil in 8411debb391d154d696fc95526bc2cecded49e23.

**Comment by ann0see** [comment]
I think the change in format can be explained by the use of another tool. 

The dmg opens fine on Catalina, which is the oldest version I have access to. @JohannesBrx has an older Mac and he should be able to test the .dmg

**Comment by hoffie** [comment]
Looks like create-dmg is hardcoded to use HFS:
https://github.com/create-dmg/create-dmg/blob/master/create-dmg#L271-L275

It also looks like HFS is the more legacy-friendly option, but might prevent us from running on M1:
https://github.com/electron-userland/electron-builder/issues/5850#issuecomment-952897519
https://github.com/jamulussoftware/jamulus/discussions/2148#discussioncomment-2123677

APFS seems to require 10.2+ (which might be fine, at least for the regular builds?).

I don't think this is a blocker for using `create-dmg`. We do not have to have M1 support just yet and we can always create a PR to `create-dmg` to permit overriding this.

**Review by hoffie — APPROVED** [review]
@JohannesBrx Thanks for the feedback.

So... as this is a requirement for moving to more recent Mac build environments and because there are no known downsides, I'd vote for merging this early for 3.9.0 so that it can be part of one of the betas.

@ann0see Thanks for working on this :)

**Comment by ann0see** [comment]
Should be merged after upcoming release (not squash merged)

**Comment by ann0see** [comment]
CHANGELOG: Build macOS .dmg files by create-dmg for building compatibility with further versions of macOS

**Comment by ann0see** [comment]
@hoffie this PR also seems to be wrongly linked in the changelog

**Comment by ann0see** [comment]
Strange. I thought we already had that in in some previous release.

**Pull request merged**
